File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int p1 = 0, p2 = 0,m = nums1.size(),n = nums2.size();
        vector<int> c(m+n);
        for(int i = 0; i < m+n; i++){
            if(p1==m) c[i] = nums2[p2++];
            else if(p2==n) c[i] = nums1[p1++];
            else if(nums1[p1]<nums2[p2]){
                c[i] = nums1[p1];
                p1++;
            }
            else{
                c[i] = nums2[p2];
                p2++;
            }
        }
        
        if (c.size()%2==1) return c[(m+n)/2];
        else return (c[(m+n-1)/2]+c[(m+n+1)/2])/2.0;
    }
};

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if(nums1.length>nums2.length) return findMedianSortedArrays(nums2,nums1);
        int m=nums1.length;
        int n=nums2.length;
        int l=0,r=m;
        while(l<=r){
            int px=l+(r-l)/2;
            int py=(m+n+1)/2-px;
            int x1=(px==0) ? Integer.MIN_VALUE:nums1[px-1];
            int x2=(py==0) ? Integer.MIN_VALUE:nums2[py-1];
            int x3=(px==m) ? Integer.MAX_VALUE:nums1[px];
            int x4=(py==n) ? Integer.MAX_VALUE:nums2[py];
            if(x1<=x4 && x2<=x3){
                if((m+n)%2==1) return Math.max(x1,x2);
                else return (Math.max(x1,x2)+Math.min(x3,x4))/2.0;
            }
            if(x1>x4) r=px-1;
            else l=px+1;
        }
        return 0;
    }
}